### PR TITLE
Rewrite `-> {}` as `Kernel.lambda {}`

### PR DESCRIPTION
**Pull request by paracycle**
Modules are not allowed to use `lambda` or `proc` in method implementations since they are actually methods on `Kernel` and the `Kernel` module is not guaranteed to be mixed-in at the module mix-in site.

However, the behaviour of `-> {}` lambda syntax is different since it does not depend on `Kernel` being available. So, the correct thing to do is to rewrite `-> {}` as `Kernel.lambda {}` to make that explicit. Which is what this PR does.

### Motivation

Resolves: #2261

### Test plan

See included automated tests. I added one explicit check for the new rewriting of the lambda syntax, and updated the expectation files in the other tests to reflect the new rewrite.

### CODE REVIEW

File: rewriter/Util.cc [hunk 1]
```diff
@@ -198,7 +198,8 @@ ast::TreePtr ASTUtil::thunkBody(core::MutableContext ctx, ast::TreePtr &node) {
     if (send->fun != core::Names::lambda() && send->fun != core::Names::proc()) {
         return nullptr;
     }
-    if (!send->recv->isSelfReference()) {
+    auto *constRecv = ast::cast_tree<ast::ConstantLit>(send->recv);
+    if (!send->recv->isSelfReference() && constRecv && constRecv->symbol != core::Symbols::Kernel()) {
```
Comment (elliottt) [inline rewriter/Util.cc:202]:
```suggestion
    if (constRecv == nullptr || constRecv->symbol != core::Symbols::Kernel()) {
```

**Comment by elliottt** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build result here:

→ https://go/builds/bui_HgmYVSkkqDeXVu

**Comment by elliottt** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build result here:

→ https://go/builds/bui_Hgmc50q4387x8z

File: rewriter/Util.cc (hunk 1 above)
Comment (paracycle) [reply rewriter/Util.cc:202]: how about we split it out into two separate conditions?

1. we check if we have a constant receiver which is not `Kernel`, which means this is not a thunk body, so return `nullptr`
2. we check if we have a receiver that is not `self`, which again means this is not a thunk body, so we return `nullptr` (this is the current behaviour)

```suggestion
    auto *constRecv = ast::cast_tree<ast::ConstantLit>(send->recv);
    if (constRecv != nullptr && constRecv->symbol != core::Symbols::Kernel()) {
        return nullptr;
    }
    if (!send->recv->isSelfReference()) {
```

This looks more correct. WDYT?

Comment (elliottt) [reply rewriter/Util.cc:202]: I like this change, thanks!

Comment (elliottt) [reply rewriter/Util.cc]: I pushed a new version of this conditional. There's now a utility function called `isKernel` that is used to test the receiver. The conditional now reads:
```cpp
if (!send->recv->isSelfReference() && !isKernel(send->recv)) {
  return nullptr;
}
```

**Comment by elliottt** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build result here:

→ https://go/builds/bui_Hgr11HuF6E09ix

**Pull request merged**